### PR TITLE
Linear jacobain calculation + linear KF unit test

### PR DESCRIPTION
**Pull request by VanillaBrooks**

### CODE REVIEW

File: src/filter/jacobian.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+    let mut l2g_slice = local_to_global_jacobian.fixed_slice_mut::<U3, U2>(0,0);
+    let rot_slice = rotation_mat.fixed_slice::<U3, U2>(0,0);
+
+    l2g_slice.copy_from(&rot_slice);
+
+    // add values into transport jacobian
+    change_mat_val!{
+        local_to_global_jacobian;
+        // [3, eT] => 1.,     // This is from the ACTS code. This might go to zero (?) since time is not being tracked
```
Comment (paulgessinger) [inline src/filter/jacobian.rs]: I'm not 100% sure what happens if we have a zero component here.

File: src/filter/angles.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
+use super::super::config::*;
+
+// TODO: Options enums for new values that are calculated: sin_phi_over_cos_theta... etc
+#[derive(PartialOrd, PartialEq, Debug)]
+pub struct Angles{
+    pub cos_theta: Real,
+    pub cos_phi: Real,
+    pub sin_theta: Real,
+    pub sin_phi: Real,
+    pub tx: Option<Real>,
```
Comment (paulgessinger) [inline src/filter/angles.rs]: Does this not involve a lot of runtime checking whether it's set? Why does this need to be optional?

File: src/filter/linear.rs [hunk 3]
```diff
@@ -3,41 +3,51 @@ use super::super::config::*;
...
+use super::utils::{SuperData, Data};
 
 #[macro_use]
 use super::macros;
 
 /// Monolithic function to handle linear KF calculations
 #[allow(dead_code)] 
 pub fn run(
+    start_location: &P3,                         // start loc used to predict initial filtered state vec
     measurement_noise_coarariance_vector: &Vec<Mat2>,  // vector of V from fruhwirth paper
```
Comment (paulgessinger) [inline src/filter/linear.rs]:
```suggestion
    measurement_noise_covarariance_vector: &Vec<Mat2>,  // vector of V from fruhwirth paper
```
This probably needs changing below.

File: src/filter/utils.rs [hunk 4]
```diff
@@ -1,17 +1,97 @@
...
+    // on the XY projection we construct, if we are in quadrants 3 or 4 then the angle
+    // generated by .angle() will be x instead of 2pi - x (returns smallest angle between vectors). 
+    // we do this here to reduce copies.
+    let adjust_factor = 
+        if y < &0. {true}
+        else {false};
+
+    let xy_projection = Vec3::new(*x, *y, 0.);
+
+    let _phi = xy_projection.angle(&x_axis);
```
Comment (paulgessinger) [inline src/filter/utils.rs:71]: Ah, what's your value range for phi? 0 < phi < 2pi? You might want consider switching to -pi < phi < pi, which you can get using `atan2(y, x)`.

File: src/filter/utils.rs [hunk 5]
```diff
@@ -27,32 +107,49 @@ pub fn vec_of_vec(num: usize) -> Vec<Vec5> {
...
                             res_vec:res_vec}
     }
-    pub fn FFI_return() {
-        unimplemented!()
-    }
 }
 
+
+
+// TODO: come up with a better name for this
```
Comment (paulgessinger) [inline src/filter/utils.rs:140]: 😆

File: tests/filter_means_gains.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+    let mut min_mat = Mat5::zeros();
+    min_mat.fill(0.00005);
+
+    print!{
+        _means_covariance_matrix,
+        _gain_cov_mat
+    }
+
+    assert!{_means_covariance_matrix - _gain_cov_mat < min_mat}
```
Comment (paulgessinger) [inline tests/filter_means_gains.rs:114]: This is pretty neat!

**Review by paulgessinger — COMMENTED** [review]
You seem pretty thorough with the tests 👍

File: src/filter/utils.rs (hunk 4 above)
Comment (VanillaBrooks) [reply src/filter/utils.rs:71]: the phi produced here is going to be 0 < phi < pi. It takes the smallest angle between the two vectors. Then we have some checks later to make sure the `_phi`  is between 0 < phi < 2pi, which is how all the other code is written. 

-pi < phi < pi could make some other things easier. I will test it out in the b-field code.

File: src/filter/linear.rs (hunk 3 above)
Comment (VanillaBrooks) [reply src/filter/linear.rs]: fixed

File: src/filter/angles.rs (hunk 2 above)
Comment (VanillaBrooks) [reply src/filter/angles.rs]: It doesnt. This was fixed in the const-b code  but i forgot to update the linear.

File: src/filter/jacobian.rs (hunk 1 above)
Comment (VanillaBrooks) [reply src/filter/jacobian.rs]: The jacobian would become non invertible, I believe.

**Pull request merged**